core/jvmMain/src/kotlinx/serialization/internal/Caching.kt

```diff
@@ -101,10 +103,37 @@ private class ConcurrentHashMapParametrizedCache<T>(private val compute: (KClass
 
 private class CacheEntry<T>(@JvmField val serializer: KSerializer<T>?)
 
+private class KTypeWrapper(val origin: KType): KType {
+    override val annotations: List<Annotation> = origin.annotations
+    override val arguments: List<KTypeProjection> = origin.arguments
+    override val classifier: KClassifier? = origin.classifier
+    override val isMarkedNullable: Boolean = origin.isMarkedNullable
+
+    override fun equals(other: Any?): Boolean {
+        if (other == null) return false
+        if (origin != other) return false
+
+        if (classifier is KClass<*>) {
```

Comment:
`classifier` can also be a `ClassReference` (https://github.com/JetBrains/kotlin/blob/6ddca4a592542ae1b800f5f04ec42d9c48464a80/libraries/stdlib/jvm/runtime/kotlin/jvm/internal/ClassReference.kt#L10) but I don't know whether it appears anywhere in real life.
Maybe @udalov knows

Reply:
`ClassReference` implements `KClass`

Reply:
`X::class` is an expression of type `KClass` which
1) if you have `kotlin-reflect.jar` in classpath, creates an object of type `KClassImpl`
2) if you don't, it creates `ClassReference`

Reply:
🤦🏻 didn't notice that it already implements KClass